src/main/scala/mesosphere/marathon/core/election/ElectionService.scala

```diff
@@ -91,10 +105,155 @@ trait ElectionCandidate {
+    *
+    * Has no effect if called before offerLeadership
+    */
+  override def abdicateLeadership(): Unit = {
+    leaderSubscription.foreach(_.cancel())
+  }
+
+  private def initializeStream(leadershipTransitionsFlow: Flow[LeadershipState, LeadershipTransition, NotUsed]) = {
+    val localEventListenerSink = Sink.foreach[LeadershipState] { event =>
+      lastState = event
+    }
+    val localTransitionSink = Sink.foreach[LeadershipTransition] { e =>
+      eventStream.publish(e)
+      if (e == LeadershipTransition.Standby) {
+        logger.error("Lost leadership; crashing")
+        crashStrategy.crash()
+      }
+    }
+
+    val (leaderStream, leaderStreamDone) =
+      RunnableGraph.fromGraph(GraphDSL.create(
+        leaderEventsSource, localEventListenerSink)(
+        (_, _)) { implicit b =>
+          { (leaderEventsSource, localEventListenerSink) =>
+            import GraphDSL.Implicits._
+            // We defensively specify eagerCancel as true; if any of the components in the stream close or fail, then
+            // we'll help to make it obvious by closing the entire graph (and, by consequence, crashing).
+            // Akka will log all stream failures, by default.
+            val stateBroadcast = b.add(Broadcast[LeadershipState](2, eagerCancel = true))
+            val transitionBroadcast = b.add(Broadcast[LeadershipTransition](2, eagerCancel = true))
+            leaderEventsSource ~> stateBroadcast.in
+            stateBroadcast ~> localEventListenerSink
+            stateBroadcast ~> leadershipTransitionsFlow ~> transitionBroadcast.in
+
+            transitionBroadcast ~> metricsSink
+            transitionBroadcast ~> localTransitionSink
+            ClosedShape
+          }
+        }).run
+
+    // When the leadership stream terminates, for any reason, we suicide
+    leaderStreamDone.onComplete {
+      case Failure(ex) =>
+        logger.info("Leadership ended with failure; exiting", ex)
+        crashStrategy.crash()
+      case Success(_) =>
+        logger.info("Leadership ended gracefully; exiting")
+        crashStrategy.crash()
+    }(ExecutionContexts.callerThread)
+
+    leaderStream
+  }
+
+  /**
+    * offerLeadership is called to candidate for leadership. It must be called by candidate only once.
+    *
+    * @param candidate is called back once elected or defeated
+    */
+  def offerLeadership(candidate: ElectionCandidate): Unit = {
+
+    /**
+      * Deduped event stream with current leader removed. Specified this way to maintain compatibility with the rest of
+      * the code base
+      */
+    val leadershipTransitionsFlow =
+      Flow[LeadershipState]
+        .map {
+          case LeadershipState.ElectedAsLeader => LeadershipTransition.ElectedAsLeader
+          case _: LeadershipState.Standby => LeadershipTransition.Standby
+        }
+        .via(EnrichedFlow.dedup(initialFilterElement = LeadershipTransition.Standby)) // if the first elements are standby, emit nothing
```

Comment:
Why `dedup`?

Reply:
If the first events are Standby, then they are removed from the stream (`initialFilterElement`).

If leadership hands off from one instance to another (both external to this instance), then while we get multiple upstream events, we only want to emit a single transition event, because our own leadership status did not change.

(in below table, empty cell indicates no event emitted; each row indicates an event emitted in time)

| Sequence | LeadershipState  | LeadershipTransition (pre-dedup) | LeadershipTransition (after dedup) |
| -------- | ---------------- | -------------------------------- | ---------------------------------- |
| 1        | Standby(None)    | Standby                          |                                    |
| 2        | Standby(host1)   | Standby                          |                                    |
| 3        | ElectedAsLeader  | ElectedAsLeader                  | ElectedAsLeader                    |
| 4        | Standby(None)    | Standby                          | Standby                            |
| 5        | Standby(host3)   | Standby                          |                                    |

Reply:
Thanks for clearing this up.